scipy/optimize/tests/test_optimize.py

```diff
@@ -1600,11 +1600,19 @@ def h(x):
         if new_cb_interface == 1:
             def callback_interface(*, intermediate_result):
                 assert intermediate_result.fun == f(intermediate_result.x)
+                if "jac" in intermediate_result.keys():
```

Comment:
I thought we had lists of the methods that support `jac` and `hess` defined somewhere like 

https://github.com/scipy/scipy/blob/46ce4bff44d1ebfe0bf1e9ea39c88b6be34defb1/scipy/optimize/_minimize.py#L40

Perhaps they are only defined somewhere in the test file. I can't look thoroughly right now, but I'd prefer to use such a list to determine whether to check the assertion because we'd also want the assertion to fall if it's missing an attribute it should have.

Reply:
`nelder-mead` for instance does not evaluate any gradients, neither do `powel` and `cg`. It took me some time to figure out that fprime, which is popping up at some places, is basically LaTeX-code. Afterwards I stepped through `_optimize.py` multiple times to find any jacobians and hessians.

I think one can also not define it on a per method basis, because in some cases the algorithms only approximate jacobians and hessians, or don't use them at all. Hence, the equality check you propose cannot be conducted purely based on the function tested alone since it depends on how `minimize` is called...

Reply:
In this test, callable gradients and Hessians are always provided.

Reply:
I added `MINIMIZE_METHODS_DERIV` and put the respective methods on there and use this list in the test.